### PR TITLE
Handle possible null symbols in `getAttributeTarget`

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1556927

I've only been able to repro/crash two of the branches, but added defensive checks for all the branches.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 1]
```diff
@@ -1291,12 +1291,12 @@ Symbol getAttributeTarget(SyntaxNode targetSyntax)
...
-                    DelegateDeclarationSyntax delegateSyntax => ((Symbols.PublicModel.NamedTypeSymbol)GetDeclaredSymbol(delegateSyntax)).UnderlyingSymbol,
+                    ParameterSyntax parameterSyntax => ((Symbols.PublicModel.ParameterSymbol)GetDeclaredSymbol(parameterSyntax))?.UnderlyingSymbol,
+                    TypeParameterSyntax typeParameterSyntax => ((Symbols.PublicModel.TypeParameterSymbol)GetDeclaredSymbol(typeParameterSyntax))?.UnderlyingSymbol,
+                    IndexerDeclarationSyntax indexerSyntax => ((Symbols.PublicModel.PropertySymbol)GetDeclaredSymbol(indexerSyntax))?.UnderlyingSymbol,
+                    AccessorDeclarationSyntax accessorSyntax => ((Symbols.PublicModel.MethodSymbol)GetDeclaredSymbol(accessorSyntax))?.UnderlyingSymbol,
+                    AnonymousFunctionExpressionSyntax anonymousFunction => ((Symbols.PublicModel.Symbol)GetSymbolInfo(anonymousFunction).Symbol)?.UnderlyingSymbol,
+                    DelegateDeclarationSyntax delegateSyntax => ((Symbols.PublicModel.NamedTypeSymbol)GetDeclaredSymbol(delegateSyntax))?.UnderlyingSymbol,
                     _ => null
                 };
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1301]: Can we replace all of this with just `GetDeclaredSymbolForNode(targetSyntax).GetSymbol()`?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs]: Good idea. Will still need different handling for lambdas (don't declare any symbols), but still a good simplification.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 2]
```diff
@@ -1289,14 +1289,14 @@ Symbol getAttributeTarget(SyntaxNode targetSyntax)
...
-                    DelegateDeclarationSyntax delegateSyntax => ((Symbols.PublicModel.NamedTypeSymbol)GetDeclaredSymbol(delegateSyntax)).UnderlyingSymbol,
+                    BaseMethodDeclarationSyntax or
+                        LocalFunctionStatementSyntax or
+                        ParameterSyntax or
+                        TypeParameterSyntax or
+                        IndexerDeclarationSyntax or
+                        AccessorDeclarationSyntax or
+                        DelegateDeclarationSyntax => GetDeclaredSymbolForNode(targetSyntax).GetSymbol(),
+                    AnonymousFunctionExpressionSyntax anonymousFunction => GetSymbolInfo(anonymousFunction).Symbol.GetSymbol(),
                     _ => null
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1300]: Any harm calling GetDeclaredSymbolForNode for whatever syntax we have?

```suggestion
                    AnonymousFunctionExpressionSyntax anonymousFunction => GetSymbolInfo(anonymousFunction).Symbol.GetSymbol(),
                    _ => GetDeclaredSymbolForNode(targetSyntax).GetSymbol()
```

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1300]: I'd rather enumerate the valid cases

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Thanks

**Review by 333fred — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Comment by Youssef1313** [comment]
I don't think yours is the same as #63567.

**Comment by AlekseyTs** [comment]
@NXTwoThou The exception looks different: "System.ArgumentException: SyntaxTree is not part of the compilation (Parameter 'syntaxTree')". This PR was fixing: "System.NullReferenceException: Object reference not set to an instance of an object."

**Pull request merged**